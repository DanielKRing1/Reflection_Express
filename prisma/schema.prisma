// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_LOCAL")
}

model User {
  password_userId             Password?           @relation(name: "password_userId")
  journal_userId              Journal[]           @relation(name: "journal_userId")
  email                       String              @id @unique

  // Set to null when delete Journal + Update id if Journal id updated
  user_lastUsedJournalId      Journal?            @relation(name: "user_lastUsedJournalId", fields: [lastUsedJId], references: [id], onDelete: SetNull, onUpdate: Cascade)
  lastUsedJId                 BigInt?             @unique
}

model Password {
  // Cascade delete/update with User
  password_userId             User                @relation(name: "password_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String              @id

  hash                        String
}

model Journal {
  user_lastUsedJournalId      User?               @relation(name: "user_lastUsedJournalId")
  inkling_journalId           Inkling[]           @relation(name: "inkling_journalId")
  journalentry_journalId      JournalEntry[]      @relation(name: "journalentry_journalId")
  thought_journalId           Thought[]           @relation(name: "thought_journalId")
  reflection_journalId        Reflection[]        @relation(name: "reflection_journalId")
  id                          BigInt              @id @default(autoincrement())
  
  // Cascade delete/update with User
  journal_userId              User                @relation(name: "journal_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String

  @@index([userId])
}

model Inkling {
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  inkling_journalId           Journal             @relation(name: "inkling_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model JournalEntry {
  reflection_journalEntryId   Reflection[]        @relation(name: "reflection_journalEntryId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  journalentry_journalId      Journal             @relation(name: "journalentry_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  @@id([journalId, timeId])
}

model Thought {
  reflection_thoughtId        Reflection[]        @relation(name: "reflection_thoughtId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  thought_journalId           Journal             @relation(name: "thought_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model Reflection {
  // Cascade delete/update with Journal
  reflection_journalId        Journal             @relation(name: "reflection_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  // Cascade delete/update with JournalEntry
  reflection_journalEntryId   JournalEntry        @relation(name: "reflection_journalEntryId", fields: [journalId, journalEntryId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  journalEntryId              DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Thought
  reflection_thoughtId        Thought             @relation(name: "reflection_thoughtId", fields: [journalId, thoughtId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  thoughtId                   DateTime            @db.Timestamptz(3)

  decision                    Int                 @db.SmallInt

  @@id([journalEntryId, thoughtId])
}
