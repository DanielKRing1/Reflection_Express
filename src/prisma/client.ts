import { PrismaClient } from "./generated/client";

let prisma = new PrismaClient({
    // log: ["query"],
});

export default prisma;
