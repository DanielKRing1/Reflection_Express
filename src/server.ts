// EXPRESS
import express, { Express } from "express";
import cors from "cors";

// MIDDLEWARE
import loggingMiddleware from "./middlewares/logging.middleware";

// ROUTES
import genRouter from "./routes";

// CONFIG
import { PORT } from "./config/server.config";

// APP
const app: Express = express();

// MIDDLEWARE
app.use(cors());
// parse requests of content-type - application/json
app.use(express.json());
// parse requests of content-type - application/x-www-form-urlencoded
app.use(
  express.urlencoded({
    extended: true,
  })
);
app.use(loggingMiddleware);

// GRAPHQL
app.use(genRouter());

// START SERVER
app.listen(PORT, () => console.log(`Server listening on port ${PORT}!`));
