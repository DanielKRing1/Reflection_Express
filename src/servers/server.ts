import genExpressApp from "./app";
import gqlServer from "./gqlServer";

import accessPromise from "../middlewares/session/access";

export default (async () => {
    const app = await genExpressApp();
    const access = await accessPromise();

    return gqlServer(app, access);
})();
