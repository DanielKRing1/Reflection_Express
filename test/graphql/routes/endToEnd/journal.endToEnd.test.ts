import http from "http";
import supertest from "supertest";

import server from "../../../../src/index";
import prisma from "../../../../src/prisma/client";

const USER_ID1 = "1";
const USER_ID2 = "2";
const PASSWORD1 = "123";
const PASSWORD2 = "456";
const JOURNAL_NAME1 = "Journal1";
const JOURNAL_NAME2 = "Journal2";
const JOURNAL_NAME3 = "Journal3";
const JOURNAL_NAME4 = "Journal4";
const JOURNAL_NAME5 = "Journal5";

describe("Journal GraphQL server", () => {
    let s: http.Server;
    let agent: supertest.SuperAgentTest;

    beforeAll(async () => {
        // Clear Database
        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;
        await prisma.$queryRaw`ALTER SEQUENCE "Journal_id_seq" RESTART WITH 1;`;

        // Create agent
        s = await server;
        agent = supertest.agent(s);

        // Create User/ Login
        await agent
            .post("/login/create-user")
            .send({ userId: USER_ID1, password: PASSWORD1 });
    });

    it("Should get no Journals (since none have been made yet)", async () => {
        const queryData = {
            query: `query Journals {
                journals {
                    id
                    userId
                    name
                }
            }`,
            variables: {},
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should create Journal 1", async () => {
        const queryData = {
            query: `mutation CreateJournal($journalName: String!) {
                createJournal(journalName: $journalName)
              }`,
            variables: { journalName: JOURNAL_NAME1 },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should get Journal 1", async () => {
        const queryData = {
            query: `query Journals {
                journals {
                    id
                    userId
                    name
                }
            }`,
            variables: {},
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should create Journal 2 and 3", async () => {
        const queryData = {
            query: `mutation CreateJournal($journalName2: String!, $journalName3: String!) {
                create2: createJournal(journalName: $journalName2)
                create3: createJournal(journalName: $journalName3)
              }`,
            variables: {
                journalName2: JOURNAL_NAME2,
                journalName3: JOURNAL_NAME3,
            },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should get Journal 1, 2, and 3", async () => {
        const queryData = {
            query: `query Journals {
                journals {
                    id
                    userId
                    name
                }
            }`,
            variables: {},
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should log in as another user", async () => {
        // "Logout"
        agent = supertest.agent(s);

        // Create new User/ Login
        await agent
            .post("/login/create-user")
            .send({ userId: USER_ID2, password: PASSWORD2 });
    });

    it("Should fail to get any Journals as User2", async () => {
        const queryData = {
            query: `query Journals {
                journals {
                    id
                    userId
                    name
                }
            }`,
            variables: {},
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should create Journal 4 and 5", async () => {
        const queryData = {
            query: `mutation CreateJournal($journalName4: String!, $journalName5: String!) {
                create4: createJournal(journalName: $journalName4)
                create5: createJournal(journalName: $journalName5)
              }`,
            variables: {
                journalName4: JOURNAL_NAME4,
                journalName5: JOURNAL_NAME5,
            },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should get Journal 4 and 5", async () => {
        const queryData = {
            query: `query Journals {
                journals {
                    id
                    userId
                    name
                }
            }`,
            variables: {},
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    afterAll(async () => {
        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;
        await prisma.$queryRaw`ALTER SEQUENCE "Journal_id_seq" RESTART WITH 1;`;

        await prisma.$disconnect();
        await s.close();
    });
});
