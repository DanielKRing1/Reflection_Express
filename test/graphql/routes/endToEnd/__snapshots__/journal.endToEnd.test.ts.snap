// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Journal GraphQL server Should create Journal 1 1`] = `
"{"data":{"createJournal":1}}
"
`;

exports[`Journal GraphQL server Should create Journal 2 and 3 1`] = `
"{"data":{"create2":1,"create3":1}}
"
`;

exports[`Journal GraphQL server Should create Journal 4 and 5 1`] = `
"{"data":{"create4":1,"create5":1}}
"
`;

exports[`Journal GraphQL server Should fail to get any Journals as User2 1`] = `
"{"data":{"journals":[]}}
"
`;

exports[`Journal GraphQL server Should get Journal 1 1`] = `
"{"data":{"journals":[{"id":"1","userId":"1","name":"Journal1"}]}}
"
`;

exports[`Journal GraphQL server Should get Journal 1, 2, and 3 1`] = `
"{"data":{"journals":[{"id":"1","userId":"1","name":"Journal1"},{"id":"2","userId":"1","name":"Journal2"},{"id":"3","userId":"1","name":"Journal3"}]}}
"
`;

exports[`Journal GraphQL server Should get Journal 4 and 5 1`] = `
"{"data":{"journals":[{"id":"4","userId":"2","name":"Journal4"},{"id":"5","userId":"2","name":"Journal5"}]}}
"
`;

exports[`Journal GraphQL server Should get no Journals (since none have been made yet) 1`] = `
"{"data":{"journals":[]}}
"
`;
